File: src/test/java/nz/co/searchwellington/controllers/LoginResourceOwnershipServiceTest.java
package nz.co.searchwellington.controllers;

import static org.mockito.Mockito.verify;
import nz.co.searchwellington.model.User;
import nz.co.searchwellington.repositories.HandTaggingService;
import nz.co.searchwellington.repositories.HibernateResourceDAO;
import nz.co.searchwellington.repositories.HibernateBackedUserDAO;

import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

public class LoginResourceOwnershipServiceTest {

	@Mock HibernateResourceDAO resourceDAO;
	@Mock HibernateBackedUserDAO userDAO;
	@Mock HandTaggingService handTaggingService;
	
	@Mock User previousOwner;
	@Mock User newOwner;
	
	private LoginResourceOwnershipService loginResourceOwnershipService;

	@Before
	public void setup() {
		MockitoAnnotations.initMocks(this);
		loginResourceOwnershipService = new LoginResourceOwnershipService(resourceDAO, userDAO, handTaggingService);
	}
	
	@Test
	public void shouldDeletePreviousUserAfterReassigningResources() throws Exception {
		loginResourceOwnershipService.reassignOwnership(previousOwner, newOwner);
		verify(userDAO).deleteUser(previousOwner);
	}
	
	@Test
	public void shouldTransferAllTaggingVotesWhenReassigningUser() throws Exception {
		loginResourceOwnershipService.reassignOwnership(previousOwner, newOwner);
		verify(handTaggingService).transferVotes(previousOwner, newOwner);
	}
	
}

File: src/test/scala/nz/co/searchwellington/controllers/LoginResourceOwnershipServiceTest.scala
package nz.co.searchwellington.controllers

import nz.co.searchwellington.model.{User, Website}
import nz.co.searchwellington.modification.ContentUpdateService
import nz.co.searchwellington.repositories.HandTaggingService
import nz.co.searchwellington.repositories.mongo.MongoRepository
import org.junit.jupiter.api.Test
import org.mockito.Mockito.{mock, verify, when}

import scala.concurrent.Future
import scala.concurrent.ExecutionContext.Implicits.global

class LoginResourceOwnershipServiceTest {
  private val mongoRepository = mock(classOf[MongoRepository])
  private val contentUpdateService =  mock(classOf[ContentUpdateService])
  private val handTaggingService =  mock(classOf[HandTaggingService])

  private val previousOwner =  mock(classOf[User])
  private val newOwner =  mock(classOf[User])

  private val resourceSubmittedAsPreviousUser = Website(owner = Some(previousOwner._id))

  private val resourcesOwnedByUser = Seq(resourceSubmittedAsPreviousUser._id)

  private val loginResourceOwnershipService = new LoginResourceOwnershipService(mongoRepository, contentUpdateService, handTaggingService)

  @Test
  def shouldDeletePreviousUserAfterReassigningResources(): Unit = {
    when(mongoRepository.getResourcesIdsOwnedBy(previousOwner)).thenReturn(Future.successful(resourcesOwnedByUser))
    when(mongoRepository.getResourceByObjectId(resourceSubmittedAsPreviousUser._id)).thenReturn(Future.successful(Some(resourceSubmittedAsPreviousUser)))
    when(contentUpdateService.update(resourceSubmittedAsPreviousUser)).thenReturn(Future.successful(true))
    when(handTaggingService.transferVotes(previousOwner, newOwner)).thenReturn(Future.successful(true))
    when(mongoRepository.removeUser(previousOwner)).thenReturn(Future.successful(true))

    loginResourceOwnershipService.reassignOwnership(previousOwner, newOwner)

    verify(mongoRepository).removeUser(previousOwner)
  }

  @Test
  def shouldTransferAllTaggingVotesWhenReassigningUser(): Unit = {
    when(mongoRepository.getResourcesIdsOwnedBy(previousOwner)).thenReturn(Future.successful(resourcesOwnedByUser))
    when(mongoRepository.getResourceByObjectId(resourceSubmittedAsPreviousUser._id)).thenReturn(Future.successful(Some(resourceSubmittedAsPreviousUser)))
    when(contentUpdateService.update(resourceSubmittedAsPreviousUser)).thenReturn(Future.successful(true))
    when(handTaggingService.transferVotes(previousOwner, newOwner)).thenReturn(Future.successful(true))
    when(mongoRepository.removeUser(previousOwner)).thenReturn(Future.successful(true))

    loginResourceOwnershipService.reassignOwnership(previousOwner, newOwner)

    verify(handTaggingService).transferVotes(previousOwner, newOwner)
  }

}